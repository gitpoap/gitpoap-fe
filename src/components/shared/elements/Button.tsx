import React from 'react';
import { Button as ButtonUI, ButtonProps } from '@mantine/core';

export const Button = (props: ButtonProps) => {
  return <ButtonUI {...props}>{props.children}</ButtonUI>;
};
