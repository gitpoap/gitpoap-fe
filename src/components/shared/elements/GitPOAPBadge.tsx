import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';

import { TextLight, MidnightBlue, ExtraHover, ExtraPressed, TextGray } from '../../../colors';

type Props = {
  className?: string;
  imgUrl: string;
  disabled?: boolean;
  size: Sizes;
  onClick?: () => void;
};

type Sizes = 'sm' | 'md' | 'lg';

type Dimensions = {
  sm: { width: number; borderSize: number };
  md: { width: number; borderSize: number };
  lg: { width: number; borderSize: number };
};

type HexProps = {
  size: Sizes;
};

const dimensions: Dimensions = {
  sm: { width: 150, borderSize: 3 },
  md: { width: 200, borderSize: 4 },
  lg: { width: 324, borderSize: 5 },
};

const HEIGHT_RATIO = 1.0467;

const Hexagon = styled.div`
  transition: 150ms background-color ease-in-out, 150ms opacity ease-in-out;
  clip-path: url(#myPath);
`;

const HexBadge = styled(Hexagon)<Props>`
  --s: ${(props) => rem(dimensions[props.size].width)};
  position: absolute;
  top: ${(props) => rem(dimensions[props.size].borderSize)};
  left: ${(props) => rem(dimensions[props.size].borderSize)};

  width: var(--s);
  height: calc(var(--s) * ${HEIGHT_RATIO});
  display: inline-block;
  font-size: initial; /* we reset the font-size if we want to add some content */
  background: url('${(props) => props.imgUrl}') no-repeat center center;
  background-size: cover;
`;

const HexOuterBorder = styled(Hexagon)<HexProps & { disabled?: boolean }>`
  position: relative;
  --s: ${(props) => rem(dimensions[props.size].width + 4 * dimensions[props.size].borderSize)};
  width: var(--s);
  height: calc(var(--s) * ${HEIGHT_RATIO});
  background-color: ${TextLight};
  cursor: pointer;

  &:hover:not([disabled]) {
    background-color: ${ExtraHover};
    ${HexBadge} {
      opacity: 0.7;
    }
  }
  &:active:not([disabled]) {
    background-color: ${ExtraPressed};
    ${HexBadge} {
      opacity: 0.5;
    }
  }
  &[disabled] {
    cursor: not-allowed;
    background-color: ${TextGray};
    ${HexBadge} {
      background: ${MidnightBlue};
    }
  }
`;

const HexInnerBorder = styled(Hexagon)<HexProps>`
  --s: ${(props) => rem(dimensions[props.size].width + 2 * dimensions[props.size].borderSize)};
  position: absolute;
  top: ${(props) => rem(dimensions[props.size].borderSize)};
  left: ${(props) => rem(dimensions[props.size].borderSize)};

  width: var(--s);
  height: calc(var(--s) * ${HEIGHT_RATIO});
  background: ${MidnightBlue};
`;

export const GitPOAPBadge = ({ className, imgUrl, disabled, size, onClick }: Props) => {
  return (
    <>
      <HexOuterBorder className={className} size={size} disabled={disabled} onClick={onClick}>
        <HexInnerBorder size={size}>
          <HexBadge imgUrl={imgUrl} size={size} />
        </HexInnerBorder>
      </HexOuterBorder>
      <svg xmlns="http://www.w3.org/2000/svg" width="0" height="0">
        <clipPath id="myPath" clipPathUnits="objectBoundingBox">
          <path d="m0.902,0.185 l-0.287,-0.157 a0.221,0.21,0,0,0,-0.214,-0.002 l-0.282,0.142 a0.221,0.21,0,0,0,-0.115,0.184 l-0.002,0.294 a0.219,0.208,0,0,0,0.112,0.181 l0.278,0.149 a0.216,0.205,0,0,0,0.207,0 l0.29,-0.149 a0.211,0.2,0,0,0,0.11,-0.174 l0.002,-0.304 a0.197,0.187,0,0,0,-0.1,-0.165" />
        </clipPath>
      </svg>
    </>
  );
};
