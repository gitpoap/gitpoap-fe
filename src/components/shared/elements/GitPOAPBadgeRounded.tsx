import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { TextLight, MidnightBlue, ExtraHover, ExtraPressed, TextGray } from '../../../colors';

type Props = {
  className?: string;
  imgUrl: string;
  disabled?: boolean;
  size: Sizes;
};

type Sizes = 'sm' | 'md' | 'lg';

type Dimensions = {
  sm: { width: number; borderSize: number };
  md: { width: number; borderSize: number };
  lg: { width: number; borderSize: number };
};

type HexProps = {
  size: Sizes;
};

const dimensions: Dimensions = {
  sm: { width: 150, borderSize: 3 },
  md: { width: 200, borderSize: 4 },
  lg: { width: 350, borderSize: 5 },
};

const Hexagon = styled.div`
  /* transition: 150ms background-color ease-in-out, 150ms opacity ease-in-out;
  clip-path: polygon(5% 25%, 50% 0, 95% 25%, 95% 75%, 50% 100%, 5% 75%); */
`;

const HexBadgeInner = styled.div`
  transform: skewX(-30deg) rotate(60deg) skewX(30deg);
  opacity: 0.8;
  transition: 150ms opacity ease;
  cursor: pointer;

  &:hover:not([disabled]) {
    opacity: 1;
  }
`;

const HexBadgeInnerContent = styled.div<{ imgUrl: string }>`
  transform: skewX(-30deg) rotate(60deg) skewX(30deg);
  background: orange;

  &:before {
    top: 0;
    height: ${rem(300)} !important;
    /* margin-top: -50px; */
    transform: skewX(-30deg) rotate(-90deg);
    background: url('${(props) => props.imgUrl}') no-repeat center center;
    background-size: cover;
    content: '';
  }
`;

const HexBadge = styled(Hexagon)<Props>`
  overflow: hidden;
  display: inline-block;
  width: 20em;
  height: 17.32em;
  transform: rotate(-30deg) skewX(30deg);
  border-radius: ${rem(6)};

  *,
  *:before {
    display: block;
    overflow: hidden;
    width: inherit;
    height: inherit;
    border-radius: inherit;
  }

  /* &::first-child ${HexBadgeInner}:before, ${HexBadgeInnerContent} {
    transform: skewX(-30deg) rotate(60deg) skewX(30deg);
    background: orange;
    content: '';
  } */
`;

// const HexOuterBorder = styled(Hexagon)<HexProps & { disabled?: boolean }>`
//   position: relative;
//   --s: ${(props) => rem(dimensions[props.size].width + 4 * dimensions[props.size].borderSize)};
//   width: var(--s);
//   height: calc(var(--s) * 1);
//   background-color: ${TextLight};
//   cursor: pointer;

//   &:hover:not([disabled]) {
//     background-color: ${ExtraHover};
//     ${HexBadge} {
//       opacity: 0.7;
//     }
//   }
//   &:active:not([disabled]) {
//     background-color: ${ExtraPressed};
//     ${HexBadge} {
//       opacity: 0.5;
//     }
//   }
//   &[disabled] {
//     cursor: not-allowed;
//     background-color: ${TextGray};
//     ${HexBadge} {
//       background: ${MidnightBlue};
//     }
//   }
// `;

// const HexInnerBorder = styled(Hexagon)<HexProps>`
//   --s: ${(props) => rem(dimensions[props.size].width + 2 * dimensions[props.size].borderSize)};
//   position: absolute;
//   top: ${(props) => rem(dimensions[props.size].borderSize)};
//   left: ${(props) => rem(dimensions[props.size].borderSize)};

//   width: var(--s);
//   height: calc(var(--s) * 1);
//   background: ${MidnightBlue};
// `;

export const GitPOAPBadgeRounded = ({ className, imgUrl, disabled, size }: Props) => {
  return (
    <HexBadge imgUrl={imgUrl} size={size}>
      <HexBadgeInner>
        <HexBadgeInnerContent imgUrl={imgUrl}></HexBadgeInnerContent>
      </HexBadgeInner>
    </HexBadge>
  );
};

/* Another option is to use SVG ~ https://stackoverflow.com/questions/67458234/mask-image-with-svg-shape-and-add-a-border*/
