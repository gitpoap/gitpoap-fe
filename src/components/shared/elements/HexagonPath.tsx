export const HexagonPath = () => {
  return (
    <svg xmlns="http://www.w3.org/2000/svg" width="0" height="0" style={{ position: 'absolute' }}>
      <clipPath id="hexagonPath" clipPathUnits="objectBoundingBox">
        <path d="M 0.5 0.0125 c -0.0277 0 -0.0553 0.0071 -0.08 0.0214 L 0.1359 0.1983 c -0.0494 0.0286 -0.08 0.0818 -0.08 0.1389 v 0.3287 c 0 0.0571 0.0307 0.1103 0.08 0.1389 l 0.2746 0.1589 c 0.0552 0.0319 0.1238 0.0319 0.179 0 l 0.2746 -0.1589 c 0.0494 -0.0286 0.08 -0.0818 0.08 -0.1389 V 0.3371 c 0 -0.0571 -0.0307 -0.1103 -0.08 -0.1389 L 0.58 0.0339 c -0.0247 -0.0143 -0.0524 -0.0214 -0.08 -0.0214 Z" />
      </clipPath>
    </svg>
  );
};
