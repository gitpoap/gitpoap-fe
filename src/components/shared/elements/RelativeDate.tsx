import { DateTime } from 'luxon';
import { TextProps, Tooltip } from '@mantine/core';
import { Text } from './Text';

type RelativeDateProps = TextProps & {
  isoDate: string;
};

export const RelativeDate = ({ isoDate, ...props }: RelativeDateProps) => {
  return (
    <Tooltip
      label={isoDate ? DateTime.fromISO(isoDate).toFormat('dd LLL yyyy HH:mm') : '-'}
      withArrow
      transition="fade"
      position="top-start"
      sx={{ textAlign: 'center' }}
    >
      <Text {...props}>{isoDate ? DateTime.fromISO(isoDate).toRelative() : ''}</Text>
    </Tooltip>
  );
};
