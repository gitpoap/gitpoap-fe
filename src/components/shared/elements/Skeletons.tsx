import React from 'react';
import { rem } from 'polished';
import { Skeleton } from '@mantine/core';
import { BackgroundPanel2, BackgroundPanel } from '../../../colors';

type Props = {
  height?: string;
};

export const POAPBadgeSkeleton = (props: Props & React.ComponentProps<typeof Skeleton>) => {
  return (
    <Skeleton
      height={props.height ?? rem(150)}
      circle
      sx={(_) => ({
        '&::after': {
          backgroundColor: BackgroundPanel,
        },
        '&::before': {
          backgroundColor: BackgroundPanel2,
        },
      })}
      {...props}
    />
  );
};
