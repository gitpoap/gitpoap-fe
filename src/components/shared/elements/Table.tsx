import { ScrollArea, Stack } from '@mantine/core';
import { rem } from 'polished';
import styled from 'styled-components';
import { BackgroundPanel } from '../../../colors';

export const TableRow = styled.tr<{ active?: boolean }>`
  cursor: pointer;
  &:hover {
    background-color: ${BackgroundPanel} !important;
  }
  ${({ active }) => active && `background-color: ${BackgroundPanel}`}
`;

type Props = {
  children: React.ReactNode;
  border?: boolean;
  headerControls?: React.ReactNode;
};

export const TableWrapper = ({ children, border = true, headerControls }: Props) => (
  <Stack
    align="center"
    justify="flex-start"
    spacing="sm"
    py={0}
    sx={{
      border: border ? `${rem(1)} solid ${BackgroundPanel}` : 'none',
      borderRadius: `${rem(6)} ${rem(6)} 0 0`,
      width: '100%',
    }}
  >
    {headerControls}
    <ScrollArea style={{ width: '100%' }}>{children}</ScrollArea>
  </Stack>
);
