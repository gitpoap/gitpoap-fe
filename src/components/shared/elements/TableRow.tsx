import styled from 'styled-components';
import { BackgroundPanel } from '../../../colors';

const TableRow = styled.tr<{ active?: boolean }>`
  cursor: pointer;
  &:hover {
    background-color: ${BackgroundPanel} !important;
  }
  ${({ active }) => active && `background-color: ${BackgroundPanel}`}
`;

export { TableRow };
