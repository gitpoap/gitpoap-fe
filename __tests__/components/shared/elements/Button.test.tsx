import { screen } from '@testing-library/react';
import 'jest-styled-components';
import { rem } from 'polished';
import { DarkGray, PrimaryBlue, TextGray, White } from '../../../../src/colors';
import { Button } from '../../../../src/components/shared/elements';
import { render } from '../../../utils';

describe('Button', () => {
  it('renders a button', () => {
    render(<Button>Button</Button>);
    const button = screen.getByRole('button', { name: 'Button' });

    expect(button).toHaveTextContent('Button');
    expect(button).toBeInTheDocument();
    expect(button).toMatchSnapshot();
  });

  it('renders a button correctly with variant - filled', () => {
    render(<Button variant="filled">Button</Button>);
    const button = screen.getByRole('button', { name: 'Button' });

    expect(button).toHaveStyle(`background-color: ${PrimaryBlue}`);
    expect(button).toHaveStyle(`color: ${White}`);
    expect(button).toMatchSnapshot();
  });

  it('renders a button correctly with variant - outlined', () => {
    render(<Button variant="outline">Button</Button>);
    const button = screen.getByRole('button', { name: 'Button' });

    expect(button).toHaveStyle(`background-color: transparent`);
    expect(button).toHaveStyle(`border-color: ${TextGray}`);
    expect(button).toHaveStyle(`border-width: ${rem(2)}`);
    expect(button).toHaveStyle(`color: ${White}`);
    expect(button).toMatchSnapshot();
  });

  it('has the correct styles when disabled', () => {
    render(
      <Button disabled variant="filled">
        Button
      </Button>,
    );

    const button = screen.getByRole('button', { name: 'Button' });

    expect(button).toBeDisabled();
    expect(button).toMatchSnapshot();

    expect(button).toHaveStyle(`background-color: ${DarkGray}`);
    expect(button).toBeDisabled();
  });
});
