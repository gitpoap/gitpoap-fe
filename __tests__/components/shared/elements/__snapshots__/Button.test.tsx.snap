// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has the correct styles when disabled 1`] = `
<button
  class="mantine-UnstyledButton-root mantine-Button-root mantine-799x9f"
  data-button="true"
  data-disabled="true"
  disabled=""
  type="button"
>
  <div
    class="mantine-3xbgk5 mantine-Button-inner"
  >
    <span
      class="mantine-qo1k2 mantine-Button-label"
      style="text-transform: uppercase;"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button renders a button 1`] = `
<button
  class="mantine-UnstyledButton-root mantine-Button-root mantine-799x9f"
  data-button="true"
  type="button"
>
  <div
    class="mantine-3xbgk5 mantine-Button-inner"
  >
    <span
      class="mantine-qo1k2 mantine-Button-label"
      style="text-transform: uppercase;"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button renders a button correctly with variant - filled 1`] = `
<button
  class="mantine-UnstyledButton-root mantine-Button-root mantine-799x9f"
  data-button="true"
  type="button"
>
  <div
    class="mantine-3xbgk5 mantine-Button-inner"
  >
    <span
      class="mantine-qo1k2 mantine-Button-label"
      style="text-transform: uppercase;"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button renders a button correctly with variant - outlined 1`] = `
<button
  class="mantine-UnstyledButton-root mantine-Button-root mantine-hyz97v"
  data-button="true"
  type="button"
>
  <div
    class="mantine-3xbgk5 mantine-Button-inner"
  >
    <span
      class="mantine-qo1k2 mantine-Button-label"
      style="text-transform: uppercase;"
    >
      Button
    </span>
  </div>
</button>
`;
